### DISCUSSION TITLE
LobeChat Synchronization: an exploration and practice of Local First | LobeChat 云同步： Local First 的一次探索与实践

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#1681] [Announcements]
> English | [简体中文](https://github.com/lobehub/lobe-chat/discussions/1681#discussioncomment-8874279)

![191shots_so](https://github.com/lobehub/lobe-chat/assets/28616219/331e02fc-36f8-469b-84b0-79facbf96c06)

# LobeChat Synchronization: an exploration and practice of Local First

in the digital age, cloud services have become the standard for almost every network application. They provide users with convenient, reliable and flexible data storage and access methods. However, with the increasing attention paid to data security and privacy issues, the concept of "Local First" has gradually become popular since last year. 

## Enlightenment of Local First 

in 2022, I read [Local-first software ](https://www.inkandswitch.com/local-first) This article advocates local as the preferred location for data processing and storage, and makes remote synchronization possible but not necessary through exquisite design. This mode of thinking has great benefits in protecting user privacy and improving application performance. The article "Local First" opened a new perspective for me, triggered deep thinking about building Local-centered products, supplemented by cloud synchronization, and also planted seeds for this practice. 

LobeChat, as an emerging application in AI Chat open source ecosystem, has a very active community. We have collected many demands for cloud synchronization. Our users expect to enjoy a coherent chat experience between different devices without sacrificing data privacy and security. From the perspective of user experience, if the spirit of Local First can be combined with seamless remote synchronization, what kind of product experience will it bring? 
This prompted me to start exploring a seamless synchronization to remote solutions in LobeChat in Local First. 
Exploration and practice

After deciding the action, we chose to build a peer-to-peer (P2P) synchronization scheme based on YJS and WebRTC technologies to try. (If you are interested, you can view: #368 for more information)

After about 2 weeks of practice, I have preliminarily completed the implementation of the whole function. This scheme has the following advantages: 
1. **traditional backend databases are not required** : With the help of YJS, a shared editing library, and WebRTC, a Web real-time communication technology, traditional databases such as Redis and MySQL are not used for data management and synchronization. 
2. **high privacy** : using WebRTC to exchange information directly between clients only needs to establish a connection through the signaling server, which not only ensures the efficiency, but also greatly enhances the security and privacy in the data transmission process. 
3. **zero conflict** : The YJS engine ensures that all participants see the same version of data status at any time and automatically resolves any conflicts that may occur. 
4. **real-time session synchronization** : After completing the whole solution, I was surprised to find that it can synchronize sessions in real time and even support multi-person online and AI sessions in the future. 

## Challenge: SoT (Source of Truth) dilemma 

of course, we also encountered challenges in the specific implementation process: how to handle operations generated during offline operations? For example, if a message is deleted from a PC and the mobile phone is not updated, what should I do if I connect to the Internet again?

Behind the concept of "Local First" lies a deeper question: where on earth is "the Source of Truth (SoT, Source of Truth)? How to determine the unique truth source when multiple independent nodes are involved and the information between the privileged nodes is inconsistent? 

At present, our implementation strategy is relatively simple. As long as the node goes online again, we try our best to achieve the latest status-to ensure that the content of each device is consistent as much as possible; New or modified operations will be broadcast to all nodes. This means that such problems can be perfectly solved only when all devices are online and can communicate. This is not ideal for the user experience, so we also believe that this function is currently only in the "experimental" state. At the same time, we very much hope to get feedback from the community and jointly discuss Local First next step in this direction. 

## Summary
To sum up, the exploration of "Local First" has brought many inspirations and changes to LobeChat. From technology selection to architecture design to problem handling methodology in specific scenarios, it has undergone reshaping. With the launch of the First version of cloud synchronization, the road to LobeChat Local First has officially begun.

**Comment by arvinxx** [comment]
> [English](https://github.com/lobehub/lobe-chat/discussions/1681#discussion-6407158) | 简体中文

![191shots_so](https://github.com/lobehub/lobe-chat/assets/28616219/331e02fc-36f8-469b-84b0-79facbf96c06)

# LobeChat 云同步： Local First 的一次探索与实践

在数字化时代，云服务几乎成为了每个网络应用的标配。它们为用户提供了便捷、可靠且灵活的数据存储和访问方式。然而，随着数据安全和隐私问题日益受到重视，“Local First”这一概念从去年开始变逐渐变得流行起来。


## Local First 的启蒙
2022年，我读到《[Local-first software](https://www.inkandswitch.com/local-first)》这篇文章，该文倡导将本地作为数据处理和存储的首选位置，并通过精巧设计使得远程同步成为可能但非必需品。这种思维模式对于保护用户隐私、提升应用性能等方面有着不小的好处。《Local First》这篇文章为我开启了新的视角，引发了对于构建以本地为中心，再辅以云同步的产品的深度思考，也为这次实践埋下了种子。

LobeChat 作为开源生态中的一个新兴 AI Chat 应用有着非常活跃的社区，我们收集到了众多关于云端同步功能的诉求。我们的用户期望在不同设备间享受连贯的聊天体验，同时不牺牲数据的私密性和安全性。从用户体验角度出发，如果能够将 Local First 精神与无缝远程同步结合起来，那么所带来的产品体验会是怎样一番景象呢？

这促使我尝试在 LobeChat 中开始探索 Local First 中一种无缝同步到远程的解决方案。

## 探索与实践
决定行动之后，我们选择了基于 YJS 和 WebRTC 技术构建点对点（P2P）同步方案进行尝试。（如果你感兴趣，可以查看：[https://github.com/lobehub/lobe-chat/discussions/368](https://github.com/lobehub/lobe-chat/discussions/368) 了解详情）

经过2周左右的实践，我初步完成了整个功能的实现，这个方案具有如下的优势：

1. **无需传统后端数据库**：借助 YJS 这一共享编辑库以及 WebRTC 这一Web实时通信技术，省去了使用 Redis、MySQL等传统数据库进行数据管理和同步。
2. **高度私密性**：利用 WebRTC 实现直接在客户端间交换信息，仅需要通过信令服务器建立连接即可，在保证效率的同时也大大增强了数据传输过程中的安全性和隐私性。
3. **零冲突**：YJS 引擎确保任何时间内所有参与者看到完全相同版本的数据状态，并自动解决可能出现的任何冲突问题。
4. **实时会话同步**：完成整个方案之后，我惊喜地发现实时的会话同步能带来「实时聊天般」的体验，甚至有潜力在未来支持多人在线同时与 AI 进行会话。


## 遭遇挑战：SoT （Source of Truth）困境

当然，在具体执行过程中我们也遇到了挑战：如何处理离线期间产生操作？比如 PC 端删除某条消息，而手机端未做更新情形下再次联网时如何处理？

**“Local First”理念背后隐藏着一个更深层次问题：“真相源头（SoT, Source of Truth）”究竟位于哪里？当涉及多个独立运行且平权节点之间信息不一致时该如何确定唯一真相来源？**

目前我们的实现策略相对简单，只要节点重新上线就尽力达成最新状态—确保各设备间内容尽量保持一致；新增或修改操作都会被广播至所有节点上。这意味着只有在所有设备均在线且可通信状态下才能完美解决此类问题。这在用户体验上不够理想，因此我们也认为该功能目前仅处于「实验」状态。同时我们非常希望获得社区的反馈，共同探讨 Local First 这个方向的下一步。

## 总结

总结而言，“Local First”的探索给 LobeChat 带来了诸多启示和改变。从技术选型到架构设计再至特定场景下问题处理方法学都经历重塑。随着第一版的云同步方案上线，我想 LobeChat 的 Local First 之旅也正式拉开帷幕。

**Reply by mwnu** [reply]
这种点对点是不是说得两个客户端同时开着才能进行同步？
考虑到它是一个网页应用，得手机端和电脑都前台打开浏览器。
如果是这样的话我感觉不如webdav方便，

**Reply by arvinxx** [reply]
> 这种点对点是不是说得两个客户端同时开着才能进行同步？
> 考虑到它是一个网页应用，得手机端和电脑都前台打开浏览器。

对

> 如果是这样的话我感觉不如 webdav 方便

嗯，毕竟不是所有人都懂怎么配置 WebDAV ，但 WebRTC 的配置门槛明显会低很多，甚至在无服务器的情况下可能是唯一方案。

后续我会看下是否可以在这套体系下接入 WebDAV / Upstash 等可以作为 SoT 的 Provider ，毕竟 WebRTC 只能算是一个实验性的 Provider，有他的局限性。

Upstash 底层用的是 Redis ，而 YJS 体系下面有 https://github.com/yjs/y-redis 可以实现 redis presistence 的接入

**Reply by mwnu** [reply]
点对点约等于手动同步，相信我基本上没有人会愿意每次都多折腾几步，而一旦忘记或者没同步到，那是相当恼火的。
WebDAV 得有一个支持webdav的服务器或云盘（自己部署的话可以在和LobeChat 同一台服务器上装个alist，多账号也有了），如果优先考虑普及性的话，那不如做一个账号系统，LibreChat就是这么做的，和LobeChat 是同类项目。

**Reply by athlan20** [reply]
老大能否接入下云盘，或者直接绕开云同步的概念，让我们手动选择不同设备的聊天记录我也完全可以接收

**Comment by linth2005** [comment]
为了便利性，牺牲一些隐私在服务端似乎也没有太多不妥，反而因为本地，带来了使用上的不便利，这种取舍不一定能给用户带来真实便利，况且，保护用户隐私也有很多手段，例如next-chat采用的upstash的Redis也同样实现了这一策略

**Reply by arvinxx** [reply]
> 为了便利性，牺牲一些隐私在服务端似乎也没有太多不妥，反而因为本地，带来了使用上的不便利，这种取舍不一定能给用户带来真实便利，

所以云同步的功能还会继续迭代，直到达到我认为的完美状态。现在只是一个初期的实验版本。

> next-chat采用的upstash的Redis也同样实现了这一策略

但根据我了解到情况，next chat 的 redis 的方案体验并没有很好，遇到冲突仍然需要用户自行解决。如果你是这个功能的深度用户，欢迎和我介绍下你的使用体验。

**Reply by linth2005** [reply]
这两个产品我都很喜欢，但是lobe-chat 在助手、插件方面，让我感受到了这个社区更多的活力，目前两个产品我都在用，就社区活跃度和版本迭代速度来看，我更倾向于使用lobe-chat。真的非常棒，在使用next-chat的过程中，我并不会保留很多的数据，因为大部分问题，问了之后是无需沉淀为知识数据的，问了解决问题，删了就行，所以你说的要用户解决冲突的问题，我基本没遇到，我通常是把upstash的数据给删了，再次重新同步，就解决了我不想要的一些废弃的数据。这方面，对于我在使用这个产品的时候，满足了在家和在公司办公的两个场景，我相信，大部分人还是在电脑上使用居多，比较还是解决工作上的一些问题，这是我在这个同步功能上的一点感受。

**Reply by arvinxx** [reply]
好的，同步这块我们会进一步研究和完善，争取早日把你说的这个场景给支持好💪

**Comment by wakefun** [comment]
请问 `signaling server`  支持在配置文件中自定义吗，目前看在[代码](https://github.com/lobehub/lobe-chat/blob/f0ec54b340286c3c1b3e8305090b5305184ce24c/src/database/core/sync.ts#L57)中是固定地址，后续有计划直接集成在项目中吗。

Could you tell me if the `signaling server` supports customization in the configuration file? Currently, it seems that a fixed address is used in the [code](https://github.com/lobehub/lobe-chat/blob/f0ec54b340286c3c1b3e8305090b5305184ce24c/src/database/core/sync.ts#L57). Are there any plans to integrate it directly into the project in the future?

**Reply by arvinxx** [reply]
信令服务器的代码在这：https://github.com/arvinxx/y-webrtc-signaling ，直接拷贝的官方示例：https://github.com/yjs/y-webrtc/blob/master/bin/server.js。 然后使用 zeabur 做的部署。

如果有需求，后续这块可以把配置放开来自己填，配置上我是有挖好槽的，只是目前没放出来。

**Comment by WyInnovate** [comment]
加油，期待
